texar/modules/encoders/xlnet_encoders.py

```diff
@@ -0,0 +1,602 @@
+            "untie_r": True,
+            "num_layers": 12,
+            "mem_len": 0,
+            "reuse_len": 0,
+            # initializer
+            "initializer": None,
+            # layer
+            "num_heads": 12,
+            "hidden_dim": 768,
+            "head_dim": 64,
+            "dropout": 0.1,
+            "attention_dropout": 0.1,
+            "use_segments": True,
+            # ffn
+            "ffn_inner_dim": 3072,
+            "activation": 'gelu',
+            # embedding
+            "vocab_size": 32000,
+            "max_seq_len": 512,
+            '@no_typecheck': ['pretrained_model_name']
+        }
+
+    def param_groups(self, lr=None, lr_layer_scale=1.0,
+                     decay_base_params=False):
+        r"""Create parameter groups for optimizers. When
+        :attr:`lr_layer_decay_rate` is not 1.0, parameters from each layer form
+        separate groups with different base learning rates.
+
+        Args:
+            lr (float): The learning rate. Can be omitted if
+                :attr:`lr_layer_decay_rate` is 1.0.
+            lr_layer_scale (float): Per-layer LR scaling rate. The `i`-th layer
+                will be scaled by `lr_layer_scale ^ (num_layers - i - 1)`.
+            decay_base_params (bool): If `True`, treat non-layer parameters
+                (e.g. embeddings) as if they're in layer 0. If `False`, these
+                parameters are not scaled.
+
+        Returns:
+            The parameter groups, used as the first argument for optimizers.
+        """
+
+        if lr_layer_scale != 1.0:
+            if lr is None:
+                raise ValueError(
+                    "lr must be specified when lr_layer_decay_rate is not 1.0")
+
+            num_layers = self._hparams.num_layers
+            scope = self.variable_scope.name
+            base_var_names = ['r_w_bias', 'r_r_bias', 'word_embedder']
+
+            if self._hparams.use_segments:
+                base_var_names.extend(['r_s_bias', 'seg_embed'])
+
+            base_params = [tf.trainable_variables(scope=scope + "/" + name)[0]
+                           for name in base_var_names]
+            base_group = {
+                "params": base_params,
+                "lr": lr * (lr_layer_scale ** num_layers
+                            if decay_base_params else 1.0)
+            }
+            param_groups = [base_group]
+            for idx in range(num_layers):
+                decay_rate = lr_layer_scale ** (num_layers - idx - 1)
+                param_group = {
+                    "params": tf.trainable_variables(
+                        scope=scope + "/" + "layer_{}".format(idx)),
+                    "lr": lr * decay_rate,
+                }
+                param_groups.append(param_group)
+        else:
+            param_groups = self.trainable_variables
+        return param_groups
+
+    @property
+    def output_size(self):
+        r"""Return the output size of the network.
```

Comment:
This doc should be more specific and clearer

Reply:
Fixed in 8a99af3